### PR TITLE
dsp/window/cmd/leakage: add utility command for parameter estimation

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Updates #1457.

### CODE REVIEW

File: dsp/window/cmd/leakage/leakage.go [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+		gammaLine,
+		deltaF0Line,
+		deltaFhalfLine,
+		spectrumLines,
+	)
+	p.Legend.Add(fmt.Sprintf("ΔF_0 = %.3v", c.deltaF0), deltaF0Line)
+	p.Legend.Add(fmt.Sprintf("ΔF_0.5 = %.3v", c.deltaFhalf), deltaFhalfLine)
+	p.Legend.Add(fmt.Sprintf("K = %.3v", c.k()), &blank)
+	p.Legend.Add(fmt.Sprintf("ɣ_max = %.3v", c.gammaMax), gammaLine)
+	p.Legend.Add(fmt.Sprintf("β = %.3v", c.beta), &blank)
```
Comment (kortschak) [inline dsp/window/cmd/leakage/leakage.go:246]: @sbinet I had difficulties getting the LaTeX equivalent of this working. when I replace these with the following and have the needed imports and setup I get a panic.
```
	p.Legend.Add(fmt.Sprintf(`${\Delta}F_0 = %.3v$`, c.deltaF0), deltaF0Line)
	p.Legend.Add(fmt.Sprintf(`${\Delta}F_{0.5} = %.3v$`, c.deltaFhalf), deltaFhalfLine)
	p.Legend.Add(fmt.Sprintf(`$K = %.3v$`, c.k()), &blank)
	p.Legend.Add(fmt.Sprintf(`${\gamma}_{max} = %.3v$`, c.gammaMax), gammaLine)
	p.Legend.Add(fmt.Sprintf(`$\beta = %.3v$`, c.beta), &blank)
```
The prefix of the panic is
```

goroutine 1 [running]:
github.com/go-latex/latex/mtex.(*visitor).Visit(0xc0001720e0, 0x7223d8, 0xc000126558, 0x0, 0x0)
	/home/daniel/pkg/mod/github.com/go-latex/latex@v0.0.0-20210118124228-b3d85cf34e07/mtex/parser.go:128 +0xec8
<snip>
```
This appears to be related to the subscript underscore.

Comment (kortschak) [reply dsp/window/cmd/leakage/leakage.go:246]: Also, the line spacing with the LaTeX handling looks pretty tight. This is probably related to the issues you filed wrt font description parameters.

Here is the non-LaTeX output

sine_20_1000_go_nolatex

and here is the LaTeX-using output (with removal of {} wrapping which swallows the text and the subscripts which panic).

sine_20_1000_go

**Comment by kortschak** [comment]
If https://golang.org/issue/39513 is implemented, the output from this tool could be included in the docs.

File: dsp/window/cmd/leakage/leakage.go [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+//  for m in mag[:len(mag)/2]:
+//  	print(m)
+//
+//  plt.plot(freq, mag[:len(mag)/2])
+//  plt.title("Spectral leakage")
+//  plt.ylabel("Amplitude (dB)")
+//  plt.xlabel("DFT bin")
+//
+//  plt.show()
+// and the be exported to leakage and compared with the Gonum
```
[the remark this answers is not in the document]
Comment (kortschak) [reply dsp/window/cmd/leakage/leakage.go]: Thanks. Fixed.

File: dsp/window/cmd/leakage/leakage.go (hunk 1 above)
Comment (sbinet) [reply dsp/window/cmd/leakage/leakage.go:249]: yeah, the handling (or lack thereof) of subscripts/superscripts is a known limitation of `go-latex`.
(that's partly whay I've started `star-tex.org/x/tex`)
- https://github.com/go-latex/latex/issues/17

as for the lack of an adequate linegap for the legend, that's news to me. (but may very well be related to https://github.com/gonum/plot/issues/676)

Comment (kortschak) [reply dsp/window/cmd/leakage/leakage.go:249]: Yeah, I wen't looking and then saw that issue. The other thing that I noticed is the `{text}` doesn't end up rendering "text". I think it should.

File: dsp/window/cmd/leakage/leakage.go [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+
+func main() {
+	name := flag.String("window", "", "specify a built-in window name")
+	param := flag.Float64("param", math.NaN(), "specify parameter for parametric windows")
+	symm := flag.Bool("symm", true, "specify whether the window is symmetric")
+	n := flag.Int("n", 20, "specify window length")
+	m := flag.Int("m", 1000, "specify sample length (must be greater than n)")
+	csv := flag.String("csv", "", "specify an input file of frequency amplitudes (dB transformed)")
+	out := flag.String("o", "", "specify output file for plot")
+	width := flag.Float64("width", 16, "specify plot width (cm)")
```
Comment (sbinet) [inline dsp/window/cmd/leakage/leakage.go:148]: perhaps use [font.ParseLength](https://pkg.go.dev/gonum.org/v1/plot@v0.9.0/font#ParseLength) to allow for nicer units handling? (_e.g.:_ `42cm`).

**Review by sbinet — APPROVED** [review]
LGTM.

perhaps add a minimal test that exercizes the `main` function? (so, probably split it into `main` and "real `main`")

File: dsp/window/cmd/leakage/leakage.go (hunk 3 above)
Comment (kortschak) [reply dsp/window/cmd/leakage/leakage.go:148]: The documentation becomes unwieldy if I do that (I need to explain what units are possible and what absent units means).

Comment (sbinet) [reply dsp/window/cmd/leakage/leakage.go:148]: ok.

**Comment by kortschak** [comment]
How confident are you with the leakage parameter estimation code? It took me a fair while to find enough information about how to calculate the values and yet they disagree with some published values (for example the values that are currently used in the documentation) and they vary with the sample and window length parameters. It would be good to have input from someone who uses DSP a lot.

It's not clear to me how I would test the program. Here also input from a DSP expert would help.

**Comment by scott-cotton** [comment]
lgtm

File: dsp/window/cmd/leakage/leakage.go [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+		t = append(t, v)
+	}
+
+	xy = make(plotter.XYs, len(t))
+	for i, a := range t {
+		if math.IsInf(a, -1) {
+			a = min
+		} else if a < min {
+			min = a
+		}
```
Comment (sbinet) [inline dsp/window/cmd/leakage/leakage.go:342]: I vaguely remember the Go style recommending to use `switch` for this case.
(feel free to ignore, though)
```suggestion
		switch {
		case math.IsInf(a, -1):
			a = min
		case if a < min:
			min = a
		}
```

**Review by sbinet — DISMISSED** [review]
LGTM (with a minor stylistic comment)

it'd be great to have a test/example or a link to a gonum.org-hosted blog post explaining and showing how this can be used.
(the package-level doc comment is great and a good starting point though.)

File: dsp/window/cmd/leakage/leakage.go (hunk 4 above)
Comment (kortschak) [reply dsp/window/cmd/leakage/leakage.go:342]: Looking more carefully at this, the logic is not correct; it works approximately right for most cases but can deviate from the intention with some inputs.

**Comment by kortschak** [comment]
Yeah, my intention is to use the comment outputs for the documentation. I could think about a blog post.

PTAL

**Review by sbinet — APPROVED** [review]

**Pull request merged**